core/src/main/java/org/geysermc/geyser/translator/inventory/LoomInventoryTranslator.java

```diff
@@ -156,16 +154,11 @@ public ItemStackResponse translateSpecialRequest(GeyserSession session, Inventor
 
         GeyserItemStack inputCopy = inventory.getItem(0).copy(1);
         inputCopy.setNetId(session.getNextItemNetId());
-        // Add the pattern manually, for better item synchronization
-        if (inputCopy.getComponents() == null) {
-            inputCopy.setComponents(new DataComponents(new HashMap<>()));
-        }
-
         BannerPatternLayer bannerPatternLayer = BannerItem.getJavaBannerPattern(session, pattern); // TODO
         if (bannerPatternLayer != null) {
-            List<BannerPatternLayer> patternsList = inputCopy.getComponents().getOrDefault(DataComponentType.BANNER_PATTERNS, new ArrayList<>());
+            List<BannerPatternLayer> patternsList = inputCopy.getComponentOrFallback(DataComponentType.BANNER_PATTERNS, new ArrayList<>());
```

Comment:
Not sure how we missed this one, but every time this is called, a new ArrayList will be created.

Reply:
Fixed in https://github.com/GeyserMC/Geyser/commit/c298061d4dea3569a285b425bc4442d9e8445adc